zerver/lib/bugdown/api_arguments_table_generator.py

```diff
@@ -0,0 +1,108 @@
+import re
+import os
+import ujson
+
+from markdown.extensions import Extension
+from markdown.preprocessors import Preprocessor
+from typing import Any, Dict, Optional, List
+import markdown
+
+REGEXP = re.compile(r'\{generate_api_arguments_table\|\s*(.+?)\s*\|\s*(.+?)\s*\}')
+
+
+class MarkdownArgumentsTableGenerator(Extension):
+    def __init__(self, configs: Optional[Dict[str, Any]]={}) -> None:
+        self.config = {
+            'base_path': ['.', 'Default location from which to evaluate relative paths for the JSON files.'],
+        }
+        for key, value in configs.items():
+            self.setConfig(key, value)
+
+    def extendMarkdown(self, md: markdown.Markdown, md_globals: Dict[str, Any]) -> None:
+        md.preprocessors.add(
+            'generate_api_arguments', APIArgumentsTablePreprocessor(md, self.getConfigs()), '_begin'
+        )
+
+
+class APIArgumentsTablePreprocessor(Preprocessor):
+    def __init__(self, md: markdown.Markdown, config: Dict[str, Any]) -> None:
+        super(APIArgumentsTablePreprocessor, self).__init__(md)
+        self.base_path = config['base_path']
+
+    def run(self, lines: List[str]) -> List[str]:
+        done = False
+        while not done:
+            for line in lines:
+                loc = lines.index(line)
+                m = REGEXP.search(line)
+
+                if m:
+                    json_filename = m.group(1)
+                    doc_filename = m.group(2)
+                    json_filename = os.path.expanduser(json_filename)
+                    if not os.path.isabs(json_filename):
+                        json_filename = os.path.normpath(
+                            os.path.join(self.base_path, json_filename)
+                        )
+                    try:
+                        with open(json_filename, 'r') as fp:
+                            json_obj = ujson.loads(fp.read())
+                            arguments = json_obj[doc_filename]
+                            text = self.render_table(arguments)
+                    except Exception as e:
+                        print('Warning: could not find file {}. Ignoring '
+                              'statement. Error: {}'.format(json_filename, e))
+                        lines[loc] = REGEXP.sub('', line)
+                        break
+
+                    line_split = REGEXP.split(line, maxsplit=0)
+                    text[0] = line_split[0] + text[0]
```

Comment:
I would add some comments here explaining what's going on, and possibly extract some local vars for clarity sake.